File: WPF/MVVM/NavigationCommand.vb
Imports System.Windows.Input

Namespace Common.MVVM

    Public Class NavigationCommand
        Inherits DependencyObject
        Implements ICommand

        Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

        Public Sub Execute(parameter As Object) Implements ICommand.Execute
            Dim KsApplication = Me.GetKsApplication()
            KsApplication.NavigateTo(Me.GetParent(), KsApplication.GetViewModel(Me.ViewType), Me.AddToStack, Me.ForceToStack)
        End Sub

        Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
            Return True
        End Function

        Private Function GetParent() As NavigationViewModel
            Dim T = Me.Parent
            If T IsNot Nothing Then
                Return T
            End If
            Dim T2 = Me.ParentType
            If T2 IsNot Nothing Then
                Return DirectCast(Me.GetKsApplication().GetViewModel(T2), NavigationViewModel)
            End If
            Return Me.GetKsApplication().DefaultNavigationView
        End Function

        Private Function GetKsApplication() As KsApplication
            Return If(Me.KsApplication, KsApplication.Current)
        End Function

#Region "AddToStack Property"
        Public Shared ReadOnly AddToStackProperty As DependencyProperty = DependencyProperty.Register("AddToStack", GetType(Boolean), GetType(NavigationCommand), New PropertyMetadata(True))

        Public Property AddToStack As Boolean
            Get
                Return DirectCast(Me.GetValue(AddToStackProperty), Boolean)
            End Get
            Set(ByVal value As Boolean)
                Me.SetValue(AddToStackProperty, value)
            End Set
        End Property
#End Region

#Region "ForceToStack Property"
        Public Shared ReadOnly ForceToStackProperty As DependencyProperty = DependencyProperty.Register("ForceToStack", GetType(Boolean), GetType(NavigationCommand), New PropertyMetadata(False))

        Public Property ForceToStack As Boolean
            Get
                Return DirectCast(Me.GetValue(ForceToStackProperty), Boolean)
            End Get
            Set(ByVal value As Boolean)
                Me.SetValue(ForceToStackProperty, value)
            End Set
        End Property
#End Region

#Region "ViewType Property"
        Public Shared ReadOnly ViewTypeProperty As DependencyProperty = DependencyProperty.Register("ViewType", GetType(Type), GetType(NavigationCommand), New PropertyMetadata(Nothing, Nothing, AddressOf ViewType_Coerce))

        Private Shared Function ViewType_Coerce(ByVal D As DependencyObject, ByVal BaseValue As Object) As Object
            Dim Value = TryCast(BaseValue, Type)

            If Value Is Nothing OrElse Not GetType(ViewModel).IsAssignableFrom(Value) Then
                Return Nothing
            End If

            Return BaseValue
        End Function

        Public Property ViewType As Type
            Get
                Return DirectCast(Me.GetValue(ViewTypeProperty), Type)
            End Get
            Set(ByVal value As Type)
                Me.SetValue(ViewTypeProperty, value)
            End Set
        End Property
#End Region

#Region "ParentType Property"
        Public Shared ReadOnly ParentTypeProperty As DependencyProperty = DependencyProperty.Register("ParentType", GetType(Type), GetType(NavigationCommand), New PropertyMetadata(Nothing, Nothing, AddressOf ParentType_Coerce))

        Private Shared Function ParentType_Coerce(ByVal D As DependencyObject, ByVal BaseValue As Object) As Object
            Dim Self = DirectCast(D, NavigationCommand)

            If Self.Parent IsNot Nothing Then
                Return Nothing
            End If

            Dim Value = TryCast(BaseValue, Type)
            If Value Is Nothing OrElse Not GetType(NavigationViewModel).IsAssignableFrom(Value) Then
                Return Nothing
            End If

            Return BaseValue
        End Function

        Public Property ParentType As Type
            Get
                Return DirectCast(Me.GetValue(ParentTypeProperty), Type)
            End Get
            Set(ByVal value As Type)
                Me.SetValue(ParentTypeProperty, value)
            End Set
        End Property
#End Region

#Region "Parent Property"
        Public Shared ReadOnly ParentProperty As DependencyProperty = DependencyProperty.Register("Parent", GetType(NavigationViewModel), GetType(NavigationCommand), New PropertyMetadata(Nothing, Nothing, AddressOf Parent_Coerce))

        Private Shared Function Parent_Coerce(ByVal D As DependencyObject, ByVal BaseValue As Object) As Object
            Dim Self = DirectCast(D, NavigationCommand)

            If Self.ParentType IsNot Nothing Then
                Return Nothing
            End If

            Return BaseValue
        End Function

        Public Property Parent As NavigationViewModel
            Get
                Return DirectCast(Me.GetValue(ParentProperty), NavigationViewModel)
            End Get
            Set(ByVal value As NavigationViewModel)
                Me.SetValue(ParentProperty, value)
            End Set
        End Property
#End Region

#Region "KsApplication Property"
        Public Shared ReadOnly KsApplicationProperty As DependencyProperty = DependencyProperty.Register("KsApplication", GetType(KsApplication), GetType(NavigationCommand), New PropertyMetadata(Nothing))

        Public Property KsApplication As KsApplication
            Get
                Return DirectCast(Me.GetValue(KsApplicationProperty), KsApplication)
            End Get
            Set(ByVal value As KsApplication)
                Me.SetValue(KsApplicationProperty, value)
            End Set
        End Property
#End Region

    End Class

End Namespace

File: WPF/MVVM/NavigationCommand.cs
using System.Windows;
using System;
using System.Windows.Input;

namespace Ks
{
    namespace Common.MVVM
    {
        public class NavigationCommand : DependencyObject, ICommand
        {
#pragma warning disable CS0067 // The event is never used
            public event EventHandler CanExecuteChanged;
#pragma warning restore CS0067 // The event is never used

            public void Execute(object parameter)
            {
                var KsApplication = this.GetKsApplication();
                KsApplication.NavigateTo(this.GetParent(), KsApplication.GetViewModel(this.ViewType), this.AddToStack, this.ForceToStack);
            }

            public bool CanExecute(object parameter)
            {
                return true;
            }

            private NavigationViewModel GetParent()
            {
                var T = this.Parent;
                if (T != null)
                    return T;
                var T2 = this.ParentType;
                if (T2 != null)
                    return (NavigationViewModel)this.GetKsApplication().GetViewModel(T2);
                return this.GetKsApplication().DefaultNavigationView;
            }

            private KsApplication GetKsApplication()
            {
                return this.KsApplication ?? KsApplication.Current;
            }

            public static readonly DependencyProperty AddToStackProperty = DependencyProperty.Register("AddToStack", typeof(bool), typeof(NavigationCommand), new PropertyMetadata(true));

            public bool AddToStack
            {
                get
                {
                    return (bool)this.GetValue(AddToStackProperty);
                }
                set
                {
                    this.SetValue(AddToStackProperty, value);
                }
            }

            public static readonly DependencyProperty ForceToStackProperty = DependencyProperty.Register("ForceToStack", typeof(bool), typeof(NavigationCommand), new PropertyMetadata(false));

            public bool ForceToStack
            {
                get
                {
                    return (bool)this.GetValue(ForceToStackProperty);
                }
                set
                {
                    this.SetValue(ForceToStackProperty, value);
                }
            }

            public static readonly DependencyProperty ViewTypeProperty = DependencyProperty.Register("ViewType", typeof(Type), typeof(NavigationCommand), new PropertyMetadata(null, null, ViewType_Coerce));

            private static object ViewType_Coerce(DependencyObject D, object BaseValue)
            {
                var Value = BaseValue as Type;

                if (Value == null || !typeof(ViewModel).IsAssignableFrom(Value))
                    return null;

                return BaseValue;
            }

            public Type ViewType
            {
                get
                {
                    return (Type)this.GetValue(ViewTypeProperty);
                }
                set
                {
                    this.SetValue(ViewTypeProperty, value);
                }
            }

            public static readonly DependencyProperty ParentTypeProperty = DependencyProperty.Register("ParentType", typeof(Type), typeof(NavigationCommand), new PropertyMetadata(null, null, ParentType_Coerce));

            private static object ParentType_Coerce(DependencyObject D, object BaseValue)
            {
                var Self = (NavigationCommand)D;

                if (Self.Parent != null)
                    return null;

                var Value = BaseValue as Type;
                if (Value == null || !typeof(NavigationViewModel).IsAssignableFrom(Value))
                    return null;

                return BaseValue;
            }

            public Type ParentType
            {
                get
                {
                    return (Type)this.GetValue(ParentTypeProperty);
                }
                set
                {
                    this.SetValue(ParentTypeProperty, value);
                }
            }

            public static readonly DependencyProperty ParentProperty = DependencyProperty.Register("Parent", typeof(NavigationViewModel), typeof(NavigationCommand), new PropertyMetadata(null, null, Parent_Coerce));

            private static object Parent_Coerce(DependencyObject D, object BaseValue)
            {
                var Self = (NavigationCommand)D;

                if (Self.ParentType != null)
                    return null;

                return BaseValue;
            }

            public NavigationViewModel Parent
            {
                get
                {
                    return (NavigationViewModel)this.GetValue(ParentProperty);
                }
                set
                {
                    this.SetValue(ParentProperty, value);
                }
            }

            public static readonly DependencyProperty KsApplicationProperty = DependencyProperty.Register("KsApplication", typeof(KsApplication), typeof(NavigationCommand), new PropertyMetadata(null));

            public KsApplication KsApplication
            {
                get
                {
                    return (KsApplication)this.GetValue(KsApplicationProperty);
                }
                set
                {
                    this.SetValue(KsApplicationProperty, value);
                }
            }
        }
    }
}
